pkg/network/prober/prober.go

```diff
@@ -48,3 +52,56 @@ func Do(ctx context.Context, target, headerValue string) (int, string, error) {
 	}
 	return resp.StatusCode, string(body), nil
 }
+
+// Done is a function that is executed after each probe returns true if the
+// probe should be finished, or false to probe again.
+type Done func(status int, body string, err error) bool
+
+// Manager manages async probes and makes sure we run concurrently only a single
+// probe for the same key.
+type Manager struct {
+	mu   sync.Mutex
+	keys sets.String
+}
+
+// New creates a new Manager.
+func New() *Manager {
+	return &Manager{
+		keys: sets.NewString(),
+	}
+}
+
+// Offer executes asynchronous probe using `target` as the key.
+// If a probe with the same key already exists, Offer will no do anything.
+// Offer starts a go routine that periodically executes `Do`, until timeout is reached.
+// After each call to `Do` the passed in `done` callback is executed that
+// should return `true` to finish probing or `false` to continue.
+// If no response from the taget is ever received or terminal condition is never reached
+// the callback is invoked with `(0, "", wait.ErrWaitTimeout)` arguments.
+func (m *Manager) Offer(ctx context.Context, target, headerValue string, done Done, period, timeout time.Duration) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if m.keys.Has(target) {
+		return
+	}
+	m.keys.Insert(target)
+	m.doAsync(ctx, target, headerValue, done, period, timeout)
```

Comment:
I'm worried that the semantics of this passing heterogeneous `done` functions is very strange. :(

I think I'd be comfortable with the set-based approach if the `done` function is setup when the `Manager` is created.

Reply:
Why? We don't depend on the callback to clean the state it is explicitly setup by us in the `doAsync`?